src/main/fault.c

```diff
@@ -0,0 +1,800 @@
+void func_80070848(Y511F0UnkStruct* arg0, s32 index) {
+    sFaultFontColor = GPACK_RGBA5551(255, 255, 255, 1);
+    Fault_Printf(32, index * 8, "%08x-%08x:%08x-%08x(%04x)", arg0->unk0, (arg0->unk0 - arg0->unk8) + arg0->unkC,
+                 arg0->unk8, arg0->unkC, arg0->unk4);
+}
+
+s32 func_800708D4(s32 arg0) {
+    Y511F0UnkStruct* var_s0 = &D_800DA840[arg0];
+    s32 j = 4;
+    s32 i = arg0;
+
+    for (j = 4, i = arg0; (i < D_800DA83C) && (j < 28); i++) {
+        if (var_s0->unk8 != 0) {
+            func_80070848(var_s0, j);
+            j++;
+        }
+        var_s0++;
+    }
+
+    return i;
+}
+
+// Overlays?
+void func_80070988(OSThread* thread) {
+    __OSThreadContext* ctx = &thread->context;
+    Y511F0UnkStruct* var_a0;
+    s32 sp44;
+    s32 var_a2;
+    s32 offset;
+    s32 var_t0;
+    Y511F0UnkStruct* var_a02;
+
+    Fault_DrawRec(22, 16, 276, 208);
+    offset = 0;
+    sp44 = -1;
+    var_t0 = 0;
+    var_a0 = &D_800DA840[0];
+    var_a2 = 0;
+    for (var_a2 = 0; var_a2 < D_800DA83C; var_a2++) {
+        if (var_a0->unk8 != 0) {
+            if ((ctx->pc >= var_a0->unk0) && (ctx->pc < ((var_a0->unk0 + var_a0->unkC) - var_a0->unk8))) {
+                offset = var_a0->unk8 - var_a0->unk0;
+                sp44 = var_a2;
+            }
+        } else {
+            var_t0++;
+        }
+        var_a0++;
+    }
+
+    sFaultFontColor = GPACK_RGBA5551(120, 255, 0, 1);
+    Fault_Printf(28, 16, "MAX:%d(%d)", D_800DA83C, var_t0);
+    Fault_Printf(95, 16, "PC:%08x", (uintptr_t)ctx->pc + offset);
+    Fault_Printf(165, 16, "RA:%08x", (uintptr_t)ctx->ra + offset);
+    Fault_Printf(235, 16, "OFF:%06x", offset);
+    if (sp44 >= 0) {
+        sFaultFontColor = GPACK_RGBA5551(255, 255, 0, 1);
+        var_a02 = &D_800DA840[sp44];
+        Fault_Printf(32, 24, "%08x-%08x:%08x-%08x(%04x)", var_a02->unk0,
+                     (var_a02->unk0 - var_a02->unk8) + var_a02->unkC, var_a02->unk8, var_a02->unkC, var_a02->unk4);
+    }
+    D_80108690 = func_800708D4(D_80108690);
+    sFaultFontColor = GPACK_RGBA5551(255, 255, 255, 1);
+}
+
+void Fault_PrintDebuggerInfo(void) {
+    Fault_DrawRec(22, 16, 276, 208);
+    sFaultFontColor = GPACK_RGBA5551(120, 120, 255, 1);
+    Fault_Printf(24, 32, "mallocRecord.allocP=%08x", mallocRecord.allocP);
+    Fault_Printf(24, 40, "mallocRecord.requestSize=%d", mallocRecord.requestSize);
+    Fault_Printf(24, 48, "mallocRecord.totalFree=%d", mallocRecord.totalFree);
+    Fault_Printf(24, 56, "mallocRecord.totalAlloc=%d", mallocRecord.totalAlloc);
+    Fault_Printf(24, 64, "mallocRecord.blockCnt=%d", mallocRecord.blockCnt);
+    Fault_Printf(24, 80, "debugger.framebuf=%08x", debugger.framebuf);
+    Fault_Printf(24, 88, "debugger.width=%d", debugger.width);
+    Fault_Printf(24, 96, "debugger.depth=%d", debugger.depth);
+    sFaultFontColor = GPACK_RGBA5551(255, 255, 255, 1);
+}
+
+void Fault_PrintStackTrace(OSThread* thread) {
+    s32 x;
+    s32 y;
+    u32* sp = (u32*)(u32)thread->context.sp;
```

Comment:
Is there a better way to handle this? I need sp to be a `u32*` to increment inline, but the double cast seems clunky.